docs/rfcs/optional-config-type.md

```diff
@@ -0,0 +1,276 @@
+# Optional[T] type for use in configuration structs
+
+## Overview
+
+In Go, types are by default set to a "zero-value", a supported value for the
+respective type that is semantically equivalent or similar to "empty", but which
+is also a valid value for the type. For many config fields, the zero value is
+not a valid configuration value and can be taken to mean that the option is
+disabled, but in certain cases it can indicate a default value, necessitating a
+way to represent the presence of a value without using a valid value for the
+type.
+
+Using standard Go without inventing any new types, the two most straightforward
+ways to accomplish this are:
+
+1. Using a separate boolean field to indicate whether the field is enabled or
+   disabled.
+2. Making the type a pointer, which makes a `nil` pointer represent that a
+   value is not present, and a valid pointer represent the desired value.
+
+Each of these approaches has deficiencies: Using a separate boolean field
+requires the user to set the boolean field to `true` in addition to setting the
+actual config option, leading to suboptimal UX. Using a pointer value has a few
+drawbacks:
+
+1. It may not be immediately obvious to a new user that a pointer type indicates
+   a field is optional.
+2. The distinction between values that are conventionally pointers (e.g. gRPC
+   configs) and optional values is lost.
+3. Setting a default value for a pointer field when decoding will set the field
+   on the resulting config struct, and additional logic must be done to unset
+   the default if the user has not specified a value.
+4. The potential for null pointer exceptions is created.
+5. Config structs are generally intended to be immutable and may be passed
+   around a lot, which makes the mutability property of pointer fields
+   an undesirable property.
```

Comment:
Drawback numbers 2 and 5 appear contradictory. Is there a reason why pointers are "conventionally" pointers, aside from mutability?
```suggestion
2. While they can be copied cheaply, pointers convey a mutable reference making
   them unsafe to share.
3. Setting a default value for a pointer field when decoding will set the field
   on the resulting config struct, and additional logic must be done to unset
   the default if the user has not specified a value.
4. The potential for null pointer exceptions is created.
```

Reply:
This may not really apply to config structs, but outside mutability, another reason to use a pointer type for a field in Go is so you don't have to pass around a large struct by value.

The specific point I was making with item 2 is that some types in popular libraries are pointers, e.g. `*zap.Logger`, `*grpc.Server`, etc. If pointers are used to indicate optional values, if you have a field that is a `*zap.Logger` type, it's difficult to tell if it's intentionally optional, or if that type is almost always used as a pointer type (`zap.Logger` as a non-pointer is rare from what I've seen). Obviously `*zap.Logger` isn't a perfect example because nobody is setting a logger in YAML, but hopefully my point still makes sense.

Reply:
@jmacd does Evan's comment solve your question?